packages/frames.js/src/middleware/neynarValidate.test.ts

```diff
@@ -0,0 +1,100 @@
+        },
+        untrustedData: {},
+      }),
+    });
+  });
+
+  it("moves to next middleware without parsing if request is not POST request", async () => {
+    const context = {
+      request: new Request("https://example.com"),
+    } as unknown as FramesContext;
+
+    const mw = neynarValidate();
+    const response = redirect("http://test.com");
+    const next = jest.fn(() => Promise.resolve(response));
+
+    await expect(mw(context, next)).resolves.toMatchObject(response);
+    expect(next).toHaveBeenCalledWith();
+  });
+
+  it("moves to next middleware if request is POST but does not have a valid JSON body", async () => {
+    const context = {
+      request: new Request("https://example.com", {
+        method: "POST",
+        body: "invalid json",
+      }),
+    } as unknown as FramesContext;
+
+    const mw = neynarValidate();
+    const response = redirect("http://test.com");
+    const next = jest.fn(() => Promise.resolve(response));
+
+    await expect(mw(context, next)).resolves.toMatchObject(response);
+    expect(next).toHaveBeenCalledWith();
+  });
+
+  it("moves to next middleware if request is POST with valid JSON but invalid body shape", async () => {
+    const context = {
+      request: new Request("https://example.com", {
+        method: "POST",
+        body: JSON.stringify({}),
+      }),
+    } as unknown as FramesContext;
+
+    const mw = neynarValidate();
+    const response = redirect("http://test.com");
+    const next = jest.fn(() => Promise.resolve(response));
+
+    await expect(mw(context, next)).resolves.toMatchObject(response);
+    expect(next).toHaveBeenCalledWith();
+  });
+
+  it("parses frame message from request body and fetches external hub context and adds it to context", async () => {
+    const context = {
+      request: sampleFrameActionRequest.clone(),
+    } as unknown as FramesContext;
+
+    const mw = neynarValidate();
+    const next = jest.fn(() => Promise.resolve(new Response()));
+
+    await mw(context, next);
+
+    expect(next).toHaveBeenCalledWith(
+      expect.objectContaining({
+        clientProtocol: { id: "farcaster", version: "vNext" },
+        message: {
+          action: {
+            cast: expect.anything() as ValidateFrameActionResponse["action"]["cast"],
+            interactor:
+              expect.anything() as ValidateFrameActionResponse["action"]["interactor"],
+            object: "validated_frame_action",
+            state: { serialized: "" },
+            tapped_button: { index: 2 },
+            timestamp: "2024-01-29T05:36:54.000Z",
+            url: "��i���j����+i̾~��z��q�Z��",
```

Comment:
Would prefer a `Buffer.from('https://some-url')` here

Reply:
not sure how that would work 🤔